### ISSUE TITLE
[Feature] Add validation set (valset) support separate from trainset

### ISSUE THREAD

**Issue by Alberto-Codes** [Alberto-Codes/gepa-adk#54]
## User Story
As a gepa-adk user, I want to separate trainset examples used for reflection from valset examples used for scoring, so that evolution generalizes and avoids overfitting.

As a user running simple experiments, I want valset to default to trainset when not provided, so that behavior remains backward compatible.

## Acceptance Criteria (Gherkin, one or more)
Scenario: Use separate trainset and valset
  Given trainset with 10 examples
  And valset with 50 examples
  When evolution runs
  Then reflective dataset uses trainset examples (capture_traces on trainset)
  And baseline and proposal scoring use valset examples
  And acceptance decisions use the valset score

Scenario: Default valset to trainset
  Given only trainset provided
  When evolution runs
  Then trainset is used for both reflection and scoring

Scenario: Candidate selector/Pareto uses valset
  Given candidate_selector is enabled
  And valset is provided
  When candidates are evaluated
  Then Pareto scores are computed from valset scores

## Technical Implementation Plan
**Tech Stack:**
- Reference upstream `gepa/api.py` and `gepa/core/engine.py` (trainset for reflection, valset for scoring/selection).

**Files:**
- src/gepa_adk/engine/async_engine.py (accept trainset + valset, split reflection vs scoring paths)
- src/gepa_adk/api.py (thread valset into engine; update logging)
- src/gepa_adk/adapters/adk_adapter.py (no API change; ensure evaluate supports both datasets)
- src/gepa_adk/domain/models.py (ensure EvolutionResult exposes valset_score from engine)
- tests/unit/engine/test_valset_scoring.py (new)
- tests/integration/engine/test_train_val_split.py (new)

**Approach:**
- Extend AsyncGEPAEngine to accept both trainset and valset (valset optional; default to trainset).
- Use trainset evaluations with capture_traces=True to build reflective datasets.
- Use valset evaluations (capture_traces=False) for baseline, proposal scoring, acceptance, and Pareto tracking.
- Cache the last trainset evaluation batch separately from the valset evaluation batch to avoid mixing reflection data with scoring data.
- Update evolve/evolve_group/evolve_workflow to pass valset into the engine and stop treating valset as post-hoc only.

## Priority
Medium - improves generalization and aligns with upstream GEPA behavior.

## Estimated Effort
Medium (3-5 days)

## Development Checklist
- [ ] All code is linted and formatted (run `pre-commit run --all-files`)
- [ ] All repository rules and conventions are followed (see CONTRIBUTING.md)
- [ ] Unit tests are added/updated as appropriate
- [ ] Integration tests are added/updated as appropriate
- [ ] All tests pass locally
- [ ] Documentation is updated if needed

## Developer Task List
- [ ] Add trainset/valset split to AsyncGEPAEngine
- [ ] Update engine state to track reflection vs scoring batches
- [ ] Update API to pass valset into engine
- [ ] Add unit tests for valset scoring
- [ ] Add integration tests for train/val separation

## Additional Context
- Current behavior: `valset` in `evolve()` is evaluated only after evolution completes (`src/gepa_adk/api.py`).
- Upstream reference in `.venv`: `gepa/core/engine.py` (valset-driven scoring and Pareto tracking).
